imagecraft/models/volume.py

```diff
@@ -0,0 +1,136 @@
+# This file is part of imagecraft.
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 3, as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Volume configuration pydantic model."""
+
+import collections
+import enum
+import re
+import typing
+import uuid
+from typing import Literal, Self, cast
+
+from craft_application.models import CraftBaseModel
+from craft_application.models.constraints import UniqueList
+from craft_parts.utils.partition_utils import VALID_PARTITION_REGEX
+from pydantic import (
+    ByteSize,
+    Field,
+    StringConstraints,
+    field_validator,
+    model_validator,
+)
+
+from imagecraft.platforms import GPT_NAME_MAX_LENGTH, FileSystem, GptType
+
+VALID_VOLUME_NAME_REGEX = VALID_PARTITION_REGEX
+VALID_STRUCTURE_NAME_REGEX = VALID_PARTITION_REGEX
+
+StructureName = typing.Annotated[
+    str,
+    StringConstraints(
+        max_length=GPT_NAME_MAX_LENGTH, pattern=VALID_STRUCTURE_NAME_REGEX
+    ),
+]
+
+VolumeName = typing.Annotated[
+    str,
+    StringConstraints(pattern=VALID_VOLUME_NAME_REGEX),
+]
+
+VOLUME_INVALID_MSG = (
+    "Volume names must only contain lowercase letters, numbers, "
+    "and hyphens, and may not begin or end with a hyphen."
+)
+
+
+class Role(str, enum.Enum):
+    """Role describes the role of given structure."""
+
+    SYSTEM_DATA = "system-data"
+    SYSTEM_BOOT = "system-boot"
+
+
+class StructureItem(CraftBaseModel):
+    """Structure item of the image."""
+
+    name: StructureName
+    id: uuid.UUID | None = None
+    role: Role
+    structure_type: GptType = Field(alias="type")
+    size: ByteSize = ByteSize(0)
```

Comment:
out of curiosity, what is the semantics of size 0?

Reply:
I was unsure about the default value, because if no size is set in the configuration imagecraft must create a partition of the size of the content that will be set in it. In the spec I mentioned "Defaults to the size of the content set in the structure." on the `size` key.

Should I set `size: ByteSize | None = None` instead?

Reply:
no, I think 0 is fine. I just haven't fully read the spec.
Might want to add a docstring to this field describing that if `size is 0`, it's computed from the contents